src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php

```diff
@@ -0,0 +1,96 @@
+<?php
+
+/*
+ * This file is part of the Sulu.
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Content\Document\Query;
+
+use Doctrine\ODM\PHPCR\Query\Builder\QueryBuilder as BaseQueryBuilder;
+use PHPCR\SessionInterface;
+use Sulu\Bundle\DocumentManagerBundle\Bridge\PropertyEncoder;
+use Sulu\Component\Content\Metadata\Factory\StructureMetadataFactory;
+use Sulu\Component\DocumentManager\DocumentStrategyInterface;
+use Sulu\Component\DocumentManager\MetadataFactoryInterface;
+use Sulu\Component\DocumentManager\Query\QueryBuilderConverter;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+
+class StructureQueryBuilderConverter extends QueryBuilderConverter
+{
+    private $structureMetaFactory;
+    private $structureMap = [];
+
+    /**
+     * @param SessionInterface $session
+     * @param EventDispatcherInterface $eventDispatcher
+     * @param MetadataFactoryInterface $metadataFactory
+     * @param PropertyEncoder $encoder
+     */
+    public function __construct(
+        SessionInterface $session,
+        EventDispatcherInterface $eventDispatcher,
+        MetadataFactoryInterface $metadataFactory,
+        PropertyEncoder $encoder,
+        DocumentStrategyInterface $strategy,
+        StructureMetadataFactory $structureMetaFactory
+    ) {
+        parent::__construct($session, $eventDispatcher, $metadataFactory, $encoder, $strategy);
+        $this->structureMetaFactory = $structureMetaFactory;
+    }
+
+    public function getQuery(BaseQueryBuilder $queryBuilder)
```

Comment:
Add comment about this (`{@inheritdoc}`), and why the below type check is necessary